### PR TITLE
feat: Expression evaluates null

### PR DESCRIPTION
**Pull request by jefersonlopeszup**
## Description

If an expression is not resolved returns null, when inside a string of multiple expressions return empty string. 

## Related Issues

#502 #537

## Tests

I added the following tests:

change ContextDataEvaluationTest

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [DCO].
- [x] All existing and new tests are passing.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [ ] No, no existing tests failed, so this is *not* a breaking change.
- [x] Yes, this is a breaking change. *If not, delete the remainder of this section.*

### CODE REVIEW

**Comment by botbeaglezup** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">The title of PR not follow the rules of: https://www.conventionalcommits.org/en/v1.0.0/</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: android/beagle/src/main/java/br/com/zup/beagle/android/components/TextInput.kt [hunk 1]
```diff
@@ -146,26 +146,30 @@ data class TextInput(
     }
 
     private fun EditText.setData(textInput: TextInput, rootView: RootView) {
```
Comment (botbeaglezup) [inline android/beagle/src/main/java/br/com/zup/beagle/android/components/TextInput.kt:148]: <table data-meta="generated_by_android">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/ZupIT/beagle/blob/cade4173d97f299ae7fb2836a2b9122a1adabe63/android/beagle/src/main/java/br/com/zup/beagle/android/components/TextInput.kt#L148"></span>The function setData appears to be too complex (16). Defined complexity threshold for methods is set to ‘15’</td>
    </tr>
  </tbody>
</table>

Comment (matheusribeirozup) [reply android/beagle/src/main/java/br/com/zup/beagle/android/components/TextInput.kt:148]: Bot is right. What do you think about splitting this function?

File: android/beagle/src/main/java/br/com/zup/beagle/android/context/ContextDataEvaluation.kt [hunk 2]
```diff
@@ -69,12 +67,8 @@ internal class ContextDataEvaluation(
         expression: String,
         bind: Bind.Expression<*>
     ) {
-        val value = evaluateExpression(contextData, bind.type, expression)
-        if (value != null) {
-            bind.evaluatedExpressions[expression] = value
-        } else {
-            bind.evaluatedExpressions.remove(expression)
-        }
+        val value = evaluateExpression(contextData, bind, expression)?:""
```
Comment (matheusribeirozup) [inline android/beagle/src/main/java/br/com/zup/beagle/android/context/ContextDataEvaluation.kt:70]: A small suggestion: add spaces between operators
```suggestion
        val value = evaluateExpression(contextData, bind, expression) ?: ""
```

File: android/beagle/src/main/java/br/com/zup/beagle/android/context/ContextDataEvaluation.kt [hunk 3]
```diff
@@ -83,20 +77,25 @@ internal class ContextDataEvaluation(
...
         val value = getValue(contextData, expression)
 
         return try {
-            if (type == String::class.java) {
+            if (bind.type == String::class.java) {
                 value?.toString()
             } else if (value is JSONArray || value is JSONObject) {
-                moshi.adapter<Any>(type).fromJson(value.toString())
-                    ?: throw IllegalStateException("JSON deserialization returned null")
+                moshi.adapter<Any>(bind.type).fromJson(value.toString())?: run {
```
Comment (matheusribeirozup) [inline android/beagle/src/main/java/br/com/zup/beagle/android/context/ContextDataEvaluation.kt:90]: The same here

**Comment by matheusribeirozup** [comment]
Please remember to keep the conventional commit standard for the PR title, as described here: https://www.conventionalcommits.org/en/v1.0.0/

File: android/beagle/src/main/java/br/com/zup/beagle/android/components/Text.kt [hunk 4]
```diff
@@ -75,13 +75,13 @@ data class Text(
 
         text.textColor?.let {
             observeBindChanges(rootView, it) { value ->
-                this.setTextColor(value)
+                value?.let { valueLet ->this.setTextColor(valueLet) }
```
Comment (uziasferreirazup) [inline android/beagle/src/main/java/br/com/zup/beagle/android/components/Text.kt:78]: the name of variable valueLet it is not a suggestive name don't answer anything, what do you think about rename to  color? 
```suggestion
                value?.let { color ->this.setTextColor(color) }
```
I think this check it is not necessary too, check if passing null the behavior it is the same.

File: android/beagle/src/main/java/br/com/zup/beagle/android/components/Text.kt [hunk 5]
```diff
@@ -75,13 +75,13 @@ data class Text(
...
             }
         }
 
         text.alignment?.let {
             observeBindChanges(rootView, it) { value ->
-                this.setAlignment(value)
+                value?.let { valueLet ->this.setAlignment(valueLet) }
```
Comment (uziasferreirazup) [inline android/beagle/src/main/java/br/com/zup/beagle/android/components/Text.kt:84]: the same comment above.

File: android/beagle/src/main/java/br/com/zup/beagle/android/components/TextInput.kt [hunk 6]
```diff
@@ -146,26 +146,30 @@ data class TextInput(
...
+        textInput.placeholder?.let { bind -> observeBindChanges(rootView, bind) { it?.let{ this.hint = it } } }
         textInput.value?.let { bind ->
-            observeBindChanges(rootView, bind) {
-                if (it != this.text.toString()) {
-                    this.removeOnTextChange()
-                    this.setText(it)
-                    this.setSelection(it.length)
-                    setUpOnTextChange(rootView)
+            observeBindChanges(rootView, bind) {valueLet->
+                valueLet?.let{
```
Comment (uziasferreirazup) [inline android/beagle/src/main/java/br/com/zup/beagle/android/components/TextInput.kt:152]: the same comment above.

File: android/beagle/src/main/java/br/com/zup/beagle/android/components/TextInput.kt [hunk 7]
```diff
@@ -146,26 +146,30 @@ data class TextInput(
...
-        textInput.placeholder?.let { bind -> observeBindChanges(rootView, bind) { this.hint = it } }
+        textInput.placeholder?.let { bind -> observeBindChanges(rootView, bind) { it?.let{ this.hint = it } } }
         textInput.value?.let { bind ->
-            observeBindChanges(rootView, bind) {
-                if (it != this.text.toString()) {
-                    this.removeOnTextChange()
-                    this.setText(it)
-                    this.setSelection(it.length)
-                    setUpOnTextChange(rootView)
+            observeBindChanges(rootView, bind) {valueLet->
```
Comment (uziasferreirazup) [inline android/beagle/src/main/java/br/com/zup/beagle/android/components/TextInput.kt:151]:
```suggestion
            observeBindChanges(rootView, bind) { text->
```

File: android/beagle/src/main/java/br/com/zup/beagle/android/components/TextInput.kt [hunk 8]
```diff
@@ -146,26 +146,30 @@ data class TextInput(
...
             }
         }
-        textInput.readOnly?.let { bind -> observeBindChanges(rootView, bind) { this.isEnabled = !it } }
-        textInput.disabled?.let { bind -> observeBindChanges(rootView, bind) { this.isEnabled = !it } }
+        textInput.readOnly?.let { bind -> observeBindChanges(rootView, bind) { it?.let{ this.isEnabled = !it } } }
+        textInput.disabled?.let { bind -> observeBindChanges(rootView, bind) { it?.let{ this.isEnabled = !it } } }
         textInput.hidden?.let { bind ->
-            observeBindChanges(rootView, bind) {
-                this.visibility = if (it) View.INVISIBLE else View.VISIBLE
+            observeBindChanges(rootView, bind) {isHiddenLet->
```
Comment (uziasferreirazup) [inline android/beagle/src/main/java/br/com/zup/beagle/android/components/TextInput.kt:165]:
```suggestion
            observeBindChanges(rootView, bind) { isHiddenLet->
```

File: android/beagle/src/main/java/br/com/zup/beagle/android/context/ContextDataEvaluation.kt [hunk 9]
```diff
@@ -83,20 +77,25 @@ internal class ContextDataEvaluation(
...
+                    BeagleMessageLogs.errorWhenExpressionEvaluateNullValue("${bind.value} : ${bind.type}")
+                    null
+                }
             } else {
-                value ?: throw IllegalStateException("Expression evaluation returned null")
+                value?: run {
```
Comment (uziasferreirazup) [inline android/beagle/src/main/java/br/com/zup/beagle/android/context/ContextDataEvaluation.kt:95]:
```suggestion
                value ?: run {
```

File: android/beagle/src/main/java/br/com/zup/beagle/android/context/ContextDataEvaluation.kt [hunk 10]
```diff
@@ -83,20 +77,25 @@ internal class ContextDataEvaluation(
             val expressionKey = it.key
             text = text.replace("@{$expressionKey}", it.value.toString())
         }
-        return text
+        return if(text.isEmpty()) null else text
     }
 
-    private fun evaluateExpression(contextData: ContextData, type: Type, expression: String): Any? {
+    private fun evaluateExpression(contextData: ContextData, bind: Bind.Expression<*>, expression: String): Any? {
```
Comment (uziasferreirazup) [inline android/beagle/src/main/java/br/com/zup/beagle/android/context/ContextDataEvaluation.kt:83]: Please add some tests to the new behavior.

File: android/beagle/src/main/java/br/com/zup/beagle/android/context/ContextDataManager.kt [hunk 11]
```diff
@@ -116,10 +116,7 @@ internal class ContextDataManager(
 
         bindings.forEach { bind ->
```
Comment (uziasferreirazup) [inline android/beagle/src/main/java/br/com/zup/beagle/android/context/ContextDataManager.kt:117]: Please add some tests to the new behavior.

**Comment by uziasferreirazup** [comment]
I recommend use auto-indent lines to don't have this problem.
Screen Shot 2020-07-20 at 14 01 15

File: android/beagle/src/main/java/br/com/zup/beagle/android/components/TextInput.kt [hunk 12]
```diff
@@ -146,26 +146,28 @@ data class TextInput(
     }
 
     private fun EditText.setData(textInput: TextInput, rootView: RootView) {
```
Comment (botbeaglezup) [inline android/beagle/src/main/java/br/com/zup/beagle/android/components/TextInput.kt:148]: <table data-meta="generated_by_android">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/ZupIT/beagle/blob/a2bd3ae026d8062c7fee8d695b3d2b7f0f956754/android/beagle/src/main/java/br/com/zup/beagle/android/components/TextInput.kt#L148"></span>The function setData appears to be too complex (15). Defined complexity threshold for methods is set to ‘15’</td>
    </tr>
  </tbody>
</table>

**Comment by uziasferreirazup** [comment]
👍 🥇

**Comment by botbeaglezup** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Big PR</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Comment by jefersonlopeszup** [comment]
this PR`s have a history commits problem, i maked a new, plis see: #608

**Pull request closed without merging**